src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -134,35 +134,66 @@ private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength,
-                    hmac.AppendData(info);
-                    counter = (byte)i;
-                    hmac.AppendData(counterSpan);
+                tempInfoBuffer = info.Length > MaxStackInfoBuffer ?
+                    (rentedTempInfoBuffer = CryptoPool.Rent(info.Length)).AsSpan(0, info.Length) :
+                    (stackalloc byte[MaxStackInfoBuffer]).Slice(0, info.Length);
 
-                    if (remainingOutput.Length >= hashLength)
-                    {
-                        t = remainingOutput.Slice(0, hashLength);
-                        remainingOutput = remainingOutput.Slice(hashLength);
-                        GetHashAndReset(hmac, t);
-                    }
-                    else
+                info.CopyTo(tempInfoBuffer);
+                infoBuffer = tempInfoBuffer;
+            }
+            else
+            {
+                infoBuffer = info;
+            }
+
+            try
+            {
+                using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, prk))
+                {
+                    for (int i = 1; ; i++)
                     {
-                        if (remainingOutput.Length > 0)
+                        hmac.AppendData(t);
+                        hmac.AppendData(infoBuffer);
+                        counter = (byte)i;
+                        hmac.AppendData(counterSpan);
+
+                        if (remainingOutput.Length >= hashLength)
                         {
-                            Debug.Assert(hashLength <= 512 / 8, "hashLength is larger than expected, consider increasing this value or using regular allocation");
-                            Span<byte> lastChunk = stackalloc byte[hashLength];
-                            GetHashAndReset(hmac, lastChunk);
-                            lastChunk.Slice(0, remainingOutput.Length).CopyTo(remainingOutput);
+                            t = remainingOutput.Slice(0, hashLength);
+                            remainingOutput = remainingOutput.Slice(hashLength);
+                            GetHashAndReset(hmac, t);
+                        }
+                        else
+                        {
+                            if (remainingOutput.Length > 0)
+                            {
+                                Debug.Assert(hashLength <= 512 / 8, "hashLength is larger than expected, consider increasing this value or using regular allocation");
+                                Span<byte> lastChunk = stackalloc byte[hashLength];
+                                GetHashAndReset(hmac, lastChunk);
+                                lastChunk.Slice(0, remainingOutput.Length).CopyTo(remainingOutput);
+                            }
+
+                            break;
                         }
-
-                        break;
                     }
                 }
             }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(tempInfoBuffer);
```

Comment:
I'm having trouble convincing myself that `info` should be given an aggressive erasure (e.g. even a stack erasure).

Once it's not worthy of stack erasure, we can get rid of the try/finally and then just only return on success.  (Clearing on rent-return is still good)

Reply:
Fair enough. Do you have thoughts on what would warrant clearing stack data? Symmetric / private keys only?

Reply:
I think my neural net maps as: temporary copies of symmetric keys, private keys, data that is going into encrypt and data that is coming out of decrypt.

There might be other cases, or more nuance.